presto-main/src/main/java/com/facebook/presto/server/protocol/QueuedStatementResource.java

```diff
@@ -175,7 +177,34 @@ public Response postStatement(
         }
 
         SessionContext sessionContext = new HttpRequestSessionContext(servletRequest, sqlParserOptions);
-        Query query = new Query(statement, sessionContext, dispatchManager, queryResultsProvider, timeoutExecutor);
+        Query query = new Query(statement, sessionContext, dispatchManager, queryResultsProvider, 0);
+        queries.put(query.getQueryId(), query);
+
+        return withCompressionConfiguration(Response.ok(query.getInitialQueryResults(uriInfo, xForwardedProto)), compressionEnabled).build();
+    }
+
+    @GET
```

Comment:
One problem here is this GET is not idempotent.

A client could encounter a transient network issue while receiving the response. This would trigger a retry on the client, which would again create a new query and increment its retry count.  At this point I'm not sure what would happen, but my guess is the retry would fail somewhere due to the query id already existing.

My thinking is we can just fix this with a quick check on the status of the returned query, but we may need to add some synchronization.